### PR TITLE
Hclang 76: Completed parser

### PR DESCRIPTION
**Pull request by C0DK**
### Description
*this is to master! that's on purpose!*

This is the completion of the parser. Everything should be working by now. The only things that doesn't work are overloading on input functions.. as in:
```
var f = (func[num] i) : num {i}
var p = (num i): num { 5 }
var p = (text i):text { "hej" }
:p f
```
this would fail as it's simply fetching the "first" function when finding the passed-function, and the first one doesn't have the correct signature. however this could easily be fixed.

### Changelog
 - Added parser
 - Added generics
 - Added symbol table
 - Added Typechecker
 - Added inline lambda

### Issues solved 
add these with the respective key as HCLANG-1 (as an example)
 - HCLANG-76

### Issues opened as a result of this 
 - HCLANG-92

## FOR REVIEWER [Do not remove]
A small checklist of things to note when you are doing a review. Make sure these things apply before going into master.
 - Did the PR actually fix the issue
 - Comment anything that doesn't make sense or anything you want clarified. Many comments are always ok
 - Make sure this is ready for exams if this is going into master
 - It's always a good idea having another reviewer look at a PR as well, if it's major. get a buddy.

### Regarding filenames
- all .tex files use pascal case
- all other files [except in /code/] use snake case

### Regarding code [if applicable]
- Check [coding conventions](https://kotlinlang.org/docs/reference/coding-conventions.html)
- Are public methods documented? 
- Are there testcases for new features / methods?
- Have you tried running the source and tested things that cannot be test-cased?

### Regarding Report [if applicable]
- Does each line have a linechange in the sourcecode?
- Does sentences make sense?
- Did you do spellchecking?
- Citations?

### CODE REVIEW

**Comment by technight** [comment]
I think develop has been merged into this branch. As a result of this some report files have been modified and is also staged for merge with master. Probably not intentional 😀

File: code/src/main/kotlin/parser/Parser.kt [hunk 1]
```diff
@@ -106,194 +97,336 @@ class Parser(val lexer: ILexer): IParser, ISymbolTable by SymbolTable(),
...
+                    accept<Token.SpecialChar.Colon>()
+                    val token = accept<Token.Identifier>()
+
+                    retrieveSymbol(token.value).handle(
+                            {
+                                if(it.any {it.containsGeneric()})
+                                    genericPassedFunctionException()
+                                AstIdentifier(token.value)
+                            },
+                            { wrongTokenTypeError("Function", token) },
```
Comment (Lambcreek) [inline code/src/main/kotlin/parser/Parser.kt]: Denne error passer vel ikke. Det er en korrekt token type (identifier), men variablen er ikke erklæret som en funktion.

File: code/src/main/kotlin/parser/symboltable/ISymbolTable.kt [hunk 2]
```diff
@@ -21,12 +21,12 @@ interface ISymbolTable{
 
     /**
      * Enters name in the symbol table's current scope.
-     * @param name The name of the identifier.
+     * @param name The name of the identfier.
```
Comment (Lambcreek) [inline code/src/main/kotlin/parser/symboltable/ISymbolTable.kt]: Det var en sjov ændring 😃

Comment (C0DK) [reply code/src/main/kotlin/parser/symboltable/ISymbolTable.kt]: den ændrer vi lige tilbage ;)

File: code/src/main/kotlin/parser/Parser.kt (hunk 1 above)
Comment (C0DK) [reply code/src/main/kotlin/parser/Parser.kt]: fixer jeg! :)

File: code/src/main/kotlin/lexer/TokenPrinter.kt [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+    is Token.Literal.Number -> token.value.toString()
+    Token.Return -> "return"
+    Token.Type.Bool -> "bool"
+    Token.Type.Func -> "func"
+    Token.Type.List -> "list"
+    Token.Type.None -> "none"
+    Token.Type.Number -> "num"
+    Token.Type.Tuple -> "tuple"
+    Token.Type.Var -> "var"
+    Token.Type.Text -> "text"
```
Comment (Abildgaard96) [inline code/src/main/kotlin/lexer/TokenPrinter.kt]: bruger vi ikke 'txt' som keyword?

File: code/src/main/kotlin/parser/typechecker/TypeChecker.kt [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+        is Expression.Value.Literal.Bool -> ExprResult.Success(AstNode.Type.Bool)
+        is Expression.Value.Literal.List ->
+            getTypeOfListExpression(expr).type.let {
+                if (it is ExprResult.Success) ExprResult.Success(AstNode.Type.List(it.type)) else it
+            }
+        is Expression.Value.Literal.Text -> ExprResult.Success(AstNode.Type.Text)
+        is Expression.Value.Literal.Tuple ->
+            ExprResult.Success(AstNode.Type.Tuple(getTypeOfTupleExpression(expr)))
+        is Expression.Value.Identifier -> retrieveSymbol(expr.name).handle(
+                /*{ it.getTypeDeclaration(listOf())?.returnType?.let { ExprResult.Success(it) }?:
```
Comment (Abildgaard96) [inline code/src/main/kotlin/parser/typechecker/TypeChecker.kt]: noget der mangler at laves, eller gammelt kode der skal slettes?

**Pull request merged**